### DISCUSSION TITLE
[이펙티브 타입스크립트] 01. 타입스크립트 알아보기 Q&A

### DISCUSSION THREAD

**Discussion by suojae** [ssuojae/thoughts-from-tech-books#1] [Ideas]
#### Rules

1. 매주 이펙티브 타입스크립트 한 챕터를 읽고 해당 챕터 범위 내에서 질문을 준비합니다.
2. 오프라인에서 만나 서로의 질문에 답해보며 타입스크립트에 대한 이해도를 높입니다.

**Comment by suojae** [comment]
### 챕터01. 타입스크립트 알아보기 문제


#### 1. 아래 코드를 타입스크립트 타입체커를 통해 타입정보를 `any`가 아닌명시적(explicit)으로 정해주고 싶다면 어떻게해야할까요? 두 가지 방법을 알려주세요 (코드를 바꾸는 방법, 설정을 바꾸는 방법)
```typescript
function add(a, b) {
    return a+b;
}
```

<br/>

#### 2. 타입스크립트로 작성된 타입과 타입스크립트가 실행되어 만들어진 자바스크립트의 타입은 항상 같다고 할  수 있나요?

<br/>

#### 3. 타입스크립트의 오버로딩이 자바스크립트에서는 어떻게 동작하는지 설명해주세요
```typescript
function add(a: number, b: number): number;
function add(a: string, b: string): string;
```

<br/>

#### 4. 타입스크립트는 `any`타입을 포함한 모든 타입에 대해서 타입체크를 하나요?

<br/>

#### 5. 덕타이핑이란 무엇인가요? 타입스크립트는 덕타이핑 언어인가요?

**Reply by startek7** [reply]
---

# 챕터 01. 타입스크립트 알아보기 문제

## 문제 1
**Q1.**  
아래 코드를 타입스크립트 타입 체커를 통해 타입 정보를 `any`가 아닌 명시적(explicit)으로 정해주고 싶다면 어떻게 해야 할까요?  
두 가지 방법을 알려주세요. (코드를 바꾸는 방법, 설정을 바꾸는 방법)

```typescript
function add(a, b) {
    return a + b;
}
```

**답변:**

### 코드를 바꾸는 방법
변수에 타입을 명시합니다:
```typescript
function add(a: number, b: number): number {
    return a + b;
}
```

### 설정을 바꾸는 방법
`tsconfig.json` 설정 파일에 아래 내용을 추가하여 `any`를 허용하지 않도록 설정합니다:
```json
{
    "compilerOptions": {
        "noImplicitAny": true
    }
}
```

---

## 문제 2
**Q2.**  
타입스크립트로 작성된 타입과 타입스크립트가 실행되어 만들어진 자바스크립트의 타입은 항상 같다고 할 수 있나요?

**답변:**  
타입스크립트가 실행되어 만들어진 자바스크립트는 **타입 정보가 제거되기 때문에** 항상 같다고 할 수는 없습니다.

---

## 문제 3
**Q3.**  
타입스크립트의 오버로딩이 자바스크립트에서는 어떻게 동작하는지 설명하세요.  

```typescript
function add(a: number, b: number): number;
function add(a: string, b: string): string;
```

**답변:**  
자바스크립트에서는 **타입 정보가 제거되기 때문에 오버로딩은 동작하지 않습니다.**  
위의 `add` 함수는 자바스크립트로 변환 시 단순히 **하나의 함수로 처리**됩니다.

---

## 문제 4
**Q4.**  
타입스크립트는 `any` 타입을 포함한 모든 타입에 대해 타입 체크를 하나요?

**답변:**  
`any` 타입은 모든 경우를 허용하는 타입이므로 타입 체크를 하지 않습니다.

---

## 문제 5
**Q5.**  
덕 타이핑(Duck Typing)이란 무엇인가요? 타입스크립트는 덕 타이핑 언어인가요?

**답변:**  
- **덕 타이핑**이란 객체가 어떤 타입에 부합하는 변수와 메서드를 가질 경우, 해당 객체를 그 타입에 속한다고 간주하는 방식입니다.  
  > "오리가 오리처럼 걷고, 오리처럼 꽥꽥댄다면, 그것은 오리다."
  
- **타입스크립트**는 덕 타이핑 언어입니다.

---

**Comment by startek7** [comment]
---

# 타입스크립트 문제와 질문

## 문제 1: 타입스크립트와 자바스크립트의 관계 이해하기
**Q1.**  
타입스크립트는 자바스크립트의 상위 집합(superset)입니다.  
이 말이 의미하는 바를 설명하고, 왜 타입스크립트는 기존 자바스크립트 파일을 그대로 지원할 수 있는지 설명하세요.

---

## 문제 2: 타입스크립트 설정 이해하기
**Q2.**  
타입스크립트 설정 파일(`tsconfig.json`)에서 `strict` 옵션이 하는 역할을 설명하세요.  
이 옵션을 활성화하면 어떤 장점이 있으며, 주의할 점은 무엇인가요?

---

## 문제 3: 코드 생성과 타입이 관계없음을 이해하기
**Q3.**  
다음 코드의 타입 오류는 컴파일 결과물에 어떤 영향을 미치나요? 예제를 보고 답하세요.

```typescript
function add(a: number, b: number): number {
  return a + b;
}

const result = add(10, "20"); // 타입 오류 발생
console.log(result);
```

---

## 문제 4: 구조적 타이핑 이해하기
**Q4.**  
구조적 타이핑(Structural Typing)과 명시적 인터페이스(Nominal Typing)의 차이를 설명하세요.  
또한, 아래 코드에서 `duck` 변수가 `Bird` 타입으로 사용될 수 있는 이유를 설명하세요.

```typescript
interface Bird {
  fly(): void;
  layEggs(): void;
}

const duck = {
  fly() { console.log("Flying!"); },
  layEggs() { console.log("Laying eggs!"); },
};

// 타입스크립트에서는 오류가 발생하지 않음
const bird: Bird = duck;
```

---

## 문제 5: any 타입 지양하기
**Q5.**  
타입스크립트에서 `any` 타입을 사용하는 것은 위험할 수 있습니다.  
`any` 타입을 사용할 때 생길 수 있는 문제점을 설명하고, 가능한 대안으로 사용할 수 있는 타입들(`unknown`, 타입 단언 등)을 예시와 함께 설명하세요.

---
